File: 5-TreeAndGraph/5.2-BinaryTreeBFS/00-bfs.js
// JS doesn't support a built-in efficient queue
//
// => Use 2 arrays:
// - a primary array (queue) for nodes of the current level
// - a secondary array (nextQueue) for nodes of the next level
// -> reassign nextQueue to queue after processing the current level
//    (avoid shifting elements when dequeue)

class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function bfs(root) {
  // Handle empty tree
  if (!root) {
    return;
  }

  // Add the root node to the primary queue
  let queue = [root];

  while (queue.length > 0) {
    // Use nextQueue to collect nodes of the next level
    const nextQueue = [];

    for (const node of queue) {
      // Process current node
      console.log(node.val);

      // Enqueue the children if they exist
      if (node.left) {
        nextQueue.push(node.left);
      }
      if (node.right) {
        nextQueue.push(node.right);
      }
    }

    // Move to the next level
    queue = nextQueue;
  }
}

File: 5-TreeAndGraph/5.2-BinaryTreeBFS/00-bfs.py
from collections import deque


class TreeNode:
    def __init__(self, val, left, right):
        self.val = val
        self.left = left
        self.right = right


def bfs(root: TreeNode):
    queue = deque([root])

    while len(queue) > 0:
        node = queue.popleft()
        print(node.val)

        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)
